enhancements/network/infiniband.md

```diff
@@ -0,0 +1,139 @@
+  - "@yrobla"
+approvers:
+  - "@knobunc"
+  - "@fepan"
+creation-date: 2020-05-06
+last-updated: 2020-05-06
+status: implementable
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# InfiniBand
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [X] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+RDMA (Remote Direct Memory Access) fabrics such as InfiniBand and RDMA over
+Converged Ethernet (RoCE) are key to unlocking scalable performance for
+demanding Artificial Intelligence (AI) and Machine Learning (ML) applications.
+RoCE v2 is supported in SR-IOV Operator as Tech Preview since 4.3, This
+enhancement is to introduce native InfiniBand support.
+
+## Motivation
+
+Enable high-performance message transmission via RDMA for distributed workloads
+such as AI/ML.
+
+### Goals
+
+- Provide native InfiniBand support on pod additional network interfaces
+
+### Non-Goals
+
+- Support RoCE v2
+- Support GPUDirect RDMA
+- Support Ethernet over InfiniBand
+- Support IP over InfiniBand (IPoIB)
+- Support InfiniBand on pod primary network interface
+
+## Proposal
+
+Extend SR-IOV Operator to support both Ethernet and InfiniBand devices.
+
+SR-IOV Hardware NIC is managed by SR-IOV Operator and will be provisioned
+with a new device type `infiniBand` or `ib`.
+
+A new Container Network Interface (CNI) plugin `ib-sriov-cni` will be added.
+
+InfiniBand support will only be available on bare metal deployment.
+
+#### Hardware
+
+The following Mellanox NICs will be supported:
+- Mellanox Technologies MT27700 Family [ConnectX-4]
+- Mellanox Technologies MT28908 Family [ConnectX-6]
+
+#### InfiniBand SR-IOV CNI
+
+InfiniBand SR-IOV CNI `ib-sriov-cni` is a 3rd party CNI plugin that moves
+SR-IOV InfiniBand device to pod namespace and configures device attributes
+such as InfiniBand GUID. It will be delivered in a separate image and copied
+to node CNI directory by a DaemonSet.
+
+#### InfiniBand Kubernetes Daemon
+
+InfiniBand Kubernetes Daemon `ib-kubernetes` is a DaemonSet that watches
+Pod object changes and manages network partitioning via InfiniBand subnet
+manager plugins.
+
+#### SR-IOV Operator
+
+SR-IOV Operator API definition will support `infiniBand` or `ib` device type
+in both `SriovNodeNetworkState` and `SriovNodeNetworkPolicy`.
+
+SR-IOV Operator will support generating net-attach-def Custom Resource with
+`ib-sriov-cni` as CNI plugin type.
+
+#### SR-IOV Config Daemon
+
+SR-IOV Config Daemon, a sub component managed by SR-IOV Operator, will be
+enhanced to change device link type to InfiniBand.
+
+#### SR-IOV Network Device Plugin
+
+SR-IOV Network Device Plugin, a sub component managed by SR-IOV Operator,
+supports selecting SR-IOV devices with link type `ether` or `infiniband`.
+
```

Comment:
im not sure where but i think we should have a section about where the infiniband SM(subnet manager) would run.
In infiniband fabric it is mandatory to have an SM running as its SDN based.

This is regardless of:

- use of `ib-kubernetes`
- use of Pkeys

the SM would need to support Virtualization, which is something that the upstream openSM does not support to my knowledge.

Reply:
I included `InfiniBand Kubernetes Daemon` in the `Proposal` section as it's needed no matter whether Pkey is configured or not.
Could you elaborate on `SM would need to support VIrtualization`?

Reply:
The SM needs to implement [infiniband virtualization](https://cw.infinibandta.org/document/dl/8126).

linux-rdma opensm implementation doesn't.
Mellanox OFED OpenSM fork does.

long-term i would expect the former to have support, but ATM its not clear to me why it was not pushed/merged upstream.

Reply:
As a note: you could get away with not using `ib-kubernetes` if Partitioning (Pkeys) is not needed by randomly assigning GUIDs to VFs via `ib-sriov-cni`.
And just manage it out-of-band (`system ctl start`).

Even when `ib-kubernetes` is used, it does not deploy UFM but rather assumes its already deployed and connects to it, So provisioning the SM is also needed.

Reply:
All InfiniBand networks must have a subnet manager running for the network to function. This is true even if two machines are connected directly with no switch involved.  From RHEL8 [doc](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_infiniband_and_rdma_networks/configuring-an-infiniband-subnet-manager_configuring-and-managing-networking)

Reply:
this is correct, but i did not understand your point.
do you mean that it is assumed that the SM is already deployed ?

Reply:
The assumption is SM will be installed and enabled via SR-IOV config daemon with its default configuration. It will be part of the SR-IOV config daemon change, for example, when `infiniband` device type is configured, it automatically start opensm service, just like when `isRdma` is configured, rdma service is started.

Reply:
I'll add it in the SR-IOV config daemon section.

Reply:
you mean SR-IOV operator ? 
you need only one per physical network.
(or several if you configure HA)

Reply:
Yes, one will make it work, several will allow active-standby mode. In this particular discussion, I was referring to enable opensm in every sriov config daemon. alternative is we could have a separate deployment for opensm service and control the number of instances available in the cluster.

Reply:
@adrianchiris does the ib-kubernetes daemon pod need to be co-located on the same node with the active opensm instance?

Reply:
i don't understand, why would you want to start opensm on every sriov node ?
it's best if we have a discussion on it on our next meeting.

> does the ib-kubernetes daemon pod need to be co-located on the same node with the active opensm instance?

This is plugin dependent.
In general, no.

Reply:
@zshi-redhat did this ever get resolved?

Reply:
@knobunc we decided to not support subnet manager in the initial infiniband release, reason being subnet manager is usually deployed by user in the infiniband switch (not managed by OpenShift).

We may add the subnet manager in the following releases if there will be a explicit asking or need to deploy it on openshift. Meanwhile, we have started the design work for including subnet manager in openshift: https://docs.google.com/document/d/1Nnb_cPLS4p-Mm71XWYbBPg7bMYhpNbBhlv_nZY49buA/edit

Reply:
The subnet manager story in 4.6 has been marked as obsolete with reason commented: https://issues.redhat.com/browse/SDN-948